RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc

```diff
@@ -346,6 +387,88 @@ void MkFitGeometryESProducer::addTECGeometry(mkfit::TrackerInfo &trk_info) {
   edm::LogVerbatim("MkFitGeometryESProducer") << ostr.str();
 }
 
+void MkFitGeometryESProducer::findRZBox(const GlobalPoint &gp, float &rmin, float &rmax, float &zmin, float &zmax) {
+  float r = gp.perp(), z = gp.z();
+  if (std::fabs(r) > rmax)
+    rmax = std::fabs(r);
+  if (std::fabs(r) < rmin)
+    rmin = std::fabs(r);
+  if (std::fabs(z) > zmax)
+    zmax = std::fabs(z);
+  if (std::fabs(z) < zmin)
+    zmin = std::fabs(z);
+}
+
+void MkFitGeometryESProducer::aggregateMaterialInfo(mkfit::TrackerInfo &trk_info) {
+  //from histogram (vector of tuples) to grid
+  for (unsigned int i = 0; i < mkfit::Config::nBinsZMat; i++) {
+    for (unsigned int j = 0; j < mkfit::Config::nBinsRMat; j++) {
+      float materialXi = 0;
+      float materialRadl = 0;
+      float sumW = 0;
+      for (auto tuple : material_histogram[i][j]) {
+        materialXi += std::get<0>(tuple) * std::get<1>(tuple);
+        materialRadl += std::get<0>(tuple) * std::get<2>(tuple);
+        sumW += std::get<0>(tuple);
+      }
+      if (sumW > 0) {
+        trk_info.material_bbxi[i][j] = materialXi / sumW;
+        trk_info.material_radl[i][j] = materialRadl / sumW;
+      } else {
+        trk_info.material_bbxi[i][j] = 0;
+        trk_info.material_radl[i][j] = 0;
+      }
+    }
+  }
+}
+
+void MkFitGeometryESProducer::fillLayers(mkfit::TrackerInfo &trk_info) {
+  for (int im = 0; im < trk_info.n_layers(); ++im) {
+    const mkfit::LayerInfo &li = trk_info.layer(im);
+    if (li.zmin() < 0 and li.zmax() < 0)
+      continue;  //neg endcap covered by pos
+    unsigned int rin, rout, zmin, zmax;
+    rin = int(abs(li.rin()) * mkfit::Config::nBinsRMat / mkfit::Config::rangeRMat);
+    rout = int(abs(li.rout()) * mkfit::Config::nBinsRMat / mkfit::Config::rangeRMat) + 1;
+    if (li.is_barrel()) {
+      zmin = 0;
+      zmax =
+          std::max(int(abs(li.zmax())), int(abs(li.zmin()))) * mkfit::Config::nBinsZMat / mkfit::Config::rangeZMat + 1;
+    } else {
+      zmin = int(abs(li.zmin()) * mkfit::Config::nBinsZMat / mkfit::Config::rangeZMat);
+      zmax = int(abs(li.zmax()) * mkfit::Config::nBinsZMat / mkfit::Config::rangeZMat) + 1;
+    }
+    for (unsigned int i = zmin; i < zmax; i++) {
+      for (unsigned int j = rin; j < rout; j++) {
+        if (trk_info.material_bbxi[i][j] == 0) {
+          float distance = 1000;
+          for (unsigned int i2 = zmin; i2 < zmax; i2++) {
+            for (unsigned int j2 = rin; j2 < rout; j2++) {
+              if (j == j2 && i == i2)
+                continue;
+              float mydist = std::hypot((float)i - i2, (float)j - j2);
```

Comment:
is the int-float conversion even needed here?
the `std::hypot` is a pretty heavy operation

```suggestion
              auto mydist2 = (i - i2)*(i - i2) + (j - j2)*(j - j2);
```
but I guess even `std::abs(i - i2) + std::abs(j - j2)` measure would work just fine